Create a shared getCodeLineAnchorGetter function

bobsilverberg opened the pull request:
Fixes #889 

The code changes are quite simple, but I found the tests confusing. I'm still not sure if the logic is actually where it needs to be in the tests, and it took me ages to figure out how to mock `ForwardComparisonMap`, which I'm now not even sure if we want/need to mock. I couldn't figure out how to mock it without making it the default export from `/pages/Compare/utils`, and the fact that we have to mock it at the top of the test file also seems strange to me.

I look forward to any and all feedback on this patch, as it's messing with my brain.

kumar303 commented on src/utils.spec.tsx:
```diff
@@ -3,13 +3,31 @@ import queryString from 'query-string';
+const mockCreateCodeLineAnchorGetter = jest
+  .fn()
+  .mockReturnValue(mockGetterFromForwardComparisonMap);
+jest.mock('./pages/Compare/utils', () => {
```
This will be very confusing for the future person who has to debug it. It suggests to me that the class is being instantiated incorrectly or that you need dependency injection or maybe something else.

bobsilverberg replied:
I've removed all the mocking.

kumar303 commented on src/utils.tsx:
```diff
@@ -61,3 +65,19 @@ export const getPathFromQueryString = (history: History) => {
+
+type GetCodeLineAnchorGetterParams = {
+  _getCodeLineAnchor?: typeof getCodeLineAnchor;
+  compareInfo?: CompareInfo | null | void;
```
It would be a mistake if the caller *omitted* this parameter and doing so would trigger incorrect behavior (without triggering any type errors). I suggest making this non-optional so that the caller is forced to do this:

```tsx
getCodeLineAnchorGetter({
  // This makes it clear that passing an undefined object was intentional.
  compareInfo: undefined
});
```

bobsilverberg replied:
Good point. This has been resolved.

kumar303 commented on src/utils.tsx:
```diff
@@ -61,3 +65,19 @@ export const getPathFromQueryString = (history: History) => {
+  compareInfo?: CompareInfo | null | void;
+};
+
+export const getCodeLineAnchorGetter = ({
```
I suggest calling this `createCodeLineAnchorGetter()` so that, at a glance, you could immediately see it's different from `getCodeLineAnchor()`

kumar303 commented on src/pages/Compare/utils.tsx:
```diff
@@ -24,7 +24,7 @@ const mergeChanges = (
 // This is a map of an old vs. new file comparison, giving preference to the
 // "forward" file, i.e. the newer file.
-export class ForwardComparisonMap {
+export default class ForwardComparisonMap {
```
Is there any reason besides the mocking problem to make this a default export? If so, I suggest changing the module name to `forwardComparisonMap.js` to avoid confusion.

bobsilverberg replied:
It was just because of the mocking. I'll change it back

kumar303 commented on src/utils.spec.tsx:
```diff
@@ -188,4 +206,38 @@ describe(__filename, () => {
+      const _getCodeLineAnchor = jest.fn();
+      expect(
+        getCodeLineAnchorGetter({ _getCodeLineAnchor, compareInfo: null }),
+      ).toEqual(_getCodeLineAnchor);
```
Can't this just be `getCodeLineAnchor`? I.E. the real function, not a mock? Since you aren't executing the function, I don't think you need to mock it.

kumar303 commented on src/utils.spec.tsx:
```diff
@@ -188,4 +206,38 @@ describe(__filename, () => {
+        getCodeLineAnchorGetter({
+          compareInfo: createFakeCompareInfo(),
+        }),
+      ).toEqual(mockGetterFromForwardComparisonMap);
```
Could you just inject a stub class, like `getCodeLineAnchorGetter({ _ForwardComparisonMap: ... })`, to solve the problem up above?

bobsilverberg replied:
I tried injecting a mock class, but couldn't get it to work. In any case I think it makes sense to remove all the mocking so that should address most of the comments in this PR.

kumar303 commented on src/pages/Compare/index.spec.tsx:
```diff
@@ -836,7 +836,9 @@ describe(__filename, () => {
     }
-    const map = new ForwardComparisonMap(diff);
+    const getterFromFactory = getCodeLineAnchorGetter({
+      compareInfo: { diff, mimeType: 'test' },
```
This looks fragile. I think you can just get a `compareInfo` object using the same mapper:

```diff
diff --git a/src/pages/Compare/index.spec.tsx b/src/pages/Compare/index.spec.tsx
index de2254f..f6b4ee9 100644
--- a/src/pages/Compare/index.spec.tsx
+++ b/src/pages/Compare/index.spec.tsx
@@ -812,11 +812,13 @@ describe(__filename, () => {
   });
 
   it('configures VersionFileViewer with a file', () => {
+    const addonId = 1;
     const baseVersionId = 1;
     const version = { ...fakeVersionWithDiff, id: baseVersionId + 1 };
     const headVersionId = version.id;
 
-    const { root } = loadDiffAndRender({
+    const { store, root } = loadDiffAndRender({
+      addonId,
       baseVersionId,
       headVersionId,
       version,
@@ -832,12 +834,13 @@ describe(__filename, () => {
       }),
     );
 
-    const diff = createInternalDiff({ baseVersionId, headVersionId, version });
-    if (!diff) {
-      throw new Error('diff was unexpectedly empty');
-    }
     const getterFromFactory = getCodeLineAnchorGetter({
-      compareInfo: { diff, mimeType: 'test' },
+      compareInfo: getCompareInfo(
+        store.getState().versions,
+        addonId,
+        baseVersionId,
+        headVersionId,
+      ),
     });
 
     const getCodeLineAnchor = viewer.prop('getCodeLineAnchor');
```

bobsilverberg replied:
Nice, thanks!

bobsilverberg commented:
I've addressed all of the issues, mostly by getting rid of all the mocking. I'm not totally sure about the last test I added for `createCodeLineAnchorGetter` because I wasn't able to figure out how to get an `equals` assertion to work, but maybe I just don't know the correct incantation. This is ready for another look.

kumar303 commented on src/utils.spec.tsx:
```diff
@@ -188,4 +195,46 @@ describe(__filename, () => {
+      expect(getterFromFactory(1)).toEqual(getterFromMap(1));
+      // And verify that a getter from the factory without compareInfo does not
+      // return the same thing.
+      expect(getterFromFactoryWithoutCompareInfo(1)).not.toEqual(
```
I think it would be fine to just check that the getter functions are different. It accomplishes the same thing and is slightly more readable to me.

```diff
diff --git a/src/utils.spec.tsx b/src/utils.spec.tsx
index 2ff167c..f56efa4 100644
--- a/src/utils.spec.tsx
+++ b/src/utils.spec.tsx
@@ -232,9 +232,7 @@ describe(__filename, () => {
       expect(getterFromFactory(1)).toEqual(getterFromMap(1));
       // And verify that a getter from the factory without compareInfo does not
       // return the same thing.
-      expect(getterFromFactoryWithoutCompareInfo(1)).not.toEqual(
-        getterFromMap(1),
-      );
+      expect(getterFromFactory).not.toEqual(getterFromFactoryWithoutCompareInfo);
     });
   });
 });
```

bobsilverberg replied:
I don't think that will work for the same reason I couldn't use an `isEqual`. Two objects from the factory are never equal to each other, so saying that `getterFromFactory` is not equal to `getterFromFactoryWithoutCompareInfo` does not prove anything. I tested this by creating a `getterFromFactory2` using the exact same call that generates `getterFromFactory`, and those two getters fail an `isEqual` assertion.

kumar303 replied:
The test I proposed is just to make sure that [this line](https://github.com/mozilla/addons-code-manager/blob/df3b815b5d07b16790696bdb31d4fda8ff96bc4c/src/utils.tsx#L76) never gets hit. I still think that would be a sufficient test here. Either way works, though.

bobsilverberg replied:
Ah, right, I was thinking about it backwards. That makes sense now. I'll change it.

kumar303 commented on src/utils.spec.tsx:
```diff
@@ -188,4 +195,46 @@ describe(__filename, () => {
+
+      // We cannot use `toEqual` for the two getters above because they are
+      // different instances, so we check that they return the same thing.
+      expect(getterFromFactory(1)).toEqual(getterFromMap(1));
```
I also tried to create a manual mock of the class object but couldn't figure it out. Oh, well. Your approach here seems fine.

bobsilverberg commented on src/utils.tsx:
```diff
@@ -61,3 +65,13 @@ export const getPathFromQueryString = (history: History) => {
 };
+
+export const createCodeLineAnchorGetter = (
+  compareInfo: CompareInfo | null | void,
```
Sure, I can do that. I guess I'm just inclined to not use an object when there's only a single argument, but the fact that it can be `undefined` is valid.
